docs/decisions/0004-openedx-filters-for-close-communication.rst

```diff
@@ -0,0 +1,45 @@
+############################################
+0004 Openedx-Filters for Close Communication
+############################################
+
+Status
+******
+
+**Draft**
+
+.. Standard statuses
+    - **Draft** if the decision is still preliminary and in experimental phase
+    - **Accepted** *(date)* once it is agreed upon
+    - **Superseded** *(date)* with a reference to its replacement if a later ADR changes or reverses the decision
+
+Context
+*******
+
+As outlined in `ADR-3 <./0003-internal-communication.rst>`__, Django signals provide excellent decoupling of components and we will be extending them to improve tracing/debugging, validation, and holistic system comprehension.
+
+There are cases for which signals are a poor fit when a component or action requires a direct response, however, we still want the source of that response data to be be decoupled and highly configurable.
+
+There is substantial cost to having multiple distinct but similar mechanisms in an ecosystem; conversely there are benefits to using a common foundation as long as basically serves the need, even if it may not fit as well as a bespoke solution.
+
```

Comment:
^ This sentence was a little confusing for me to parse. If I'm understanding your meaning correctly, I think this may be a simpler restatement:

Signals and Filters are similar-yet-distinct mechanisms. Using both in the same ecosystem (as opposed to only signals) has a potential confusion cost, but it also gives benefit as long as each serves their purpose well.

Reply:
Thanks, I will try to clarify; I mean to make the case that creating our own version of filters would have costs and using an existing implementation would have benefits

Reply:
Re-ordered a bit and tried to clarify my intention, let me know if it's clearer.

Reply:
Yes to the reordering, that makes it easier to understand! I actually was all wrong about what you meant in the last sentence: I thought you were comparing using signals & filters vs signals for everything. I had no idea that your last sentence was referring to signals & filters vs signals plus creating our own version of filters(!). (I thought the 'bespoke' bit was choosing to use both, but you meant actually write it). 

To make sure I'm understanding, is this an accurate restatement of what you're saying?
There is substantial cost to having multiple distinct but similar mechanisms in an ecosystem (both signals and filters); but these provide the benefit of having a common foundation as long as each part serves its need. The alternative would be writing a bespoke version of filters which could fit better for us, but that would add time, effort and complexity. Therefore we're choosing to use signals and filters as they're already defined.

Reply:
Tried to clarify a bit more; the cost I was referring to in this paragraph is _only_ the cost of creating a bespoke solution that is similar to filters but different.

Should I also add a paragraph addressing the cost of having two mechanisms for components to communicate?

Reply:
Dude this is so much clearer! love it! And now I think the first parts are so clear that we can/should just remove the last paragraph, right? Like, if we're *not* going to make a bespoke solution, will it be necessary to say anything about not doing it?

Reply:
Dropped the bit about not building a bespoke solution, but kept the part about benefits of using a shared foundation.